### PR TITLE
[TT-11505 ] OAS controller reconciliation logic optimisations

### PR DESCRIPTION
**Pull request by buraksekili**
## Description

- Update delete reconciliation logic to handle deletion of non-existing APIs.
- Optimise reconcilition of OAS APIs - prevent sending redundant PUT requests to Tyk which may be costly.

## Related Issue

https://tyktech.atlassian.net/browse/TT-11505

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If PRing from your fork, don't come from your `master`!
- [ ] Make sure you are making a pull request against our **`master` branch** (left side). Also, it would be best if you started *your change* off *our latest `master`*.
- [ ] Make sure you are updating [CHANGELOG.md](https://github.com/TykTechnologies/tyk-operator/blob/master/CHANGELOG.md) based on your changes.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] If you've changed API models, please update CRDs.
  - [ ] `make manifests`
  - [ ] `make helm`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `gofmt -s -w .`
  - [ ] `go vet ./...`
  - [ ] `golangci-lint run`

### CODE REVIEW

**Review by komalsukhani — CHANGES_REQUESTED** [review]

File: api/v1alpha1/tykoasapidefinition_types.go [hunk 1]
```diff
@@ -75,6 +75,21 @@ type TykOasApiDefinitionStatus struct {
 	LatestTransaction TransactionInfo `json:"latestTransaction,omitempty"`
 	// IngressTemplate shows whether this CR is used as Ingress Template or not.
 	IngressTemplate bool `json:"ingressTemplate,omitempty"`
+
+	// LatestTykSpecHash stores the hash of OAS API Definition created on Tyk. This information is updated after
+	// creating or updating the TykOasApiDefinition. It is useful for Operator to understand the need for
+	// running update operation or not. If there is a change in latestTykSpecHash as well as latestCRDSpecHash,
+	// Operator runs update logic and updates resources on Tyk Gateway or Tyk Dashboard.
+	LatestTykSpecHash string `json:"latestTykSpecHash,omitempty"`
```
[the remark this answers is not in the document]
Comment (buraksekili) [reply api/v1alpha1/tykoasapidefinition_types.go:83]: Actually, I used LatestTykSpecHash to make controllers consistent with each other as the api_definition controller also uses the same name. Do you think it is not relevant to the intention of the field? if so, I can rename it.

File: controllers/tykoasapidefinition_controller.go [hunk 2]
```diff
@@ -230,23 +236,35 @@ func (r *TykOasApiDefinitionReconciler) createOrUpdateTykOASAPI(ctx context.Cont
 
 	if tykOASDoc, err = r.processClientCertificate(ctx, env, log, tykOASCrd, tykOASDoc); err != nil {
 		log.Error(err, "failed to process client certificate")
-		return "", err
+		return "", "", err
 	}
 
+	cmHash := calculateHash(tykOASDoc)
```
[the remark this answers is not in the document]
Comment (buraksekili) [reply controllers/tykoasapidefinition_controller.go:242]: You are right, i will move `cmHash` initialization right after fetching ConfigMap.

File: controllers/tykoasapidefinition_controller.go [hunk 3]
```diff
@@ -230,23 +236,35 @@ func (r *TykOasApiDefinitionReconciler) createOrUpdateTykOASAPI(ctx context.Cont
...
 		if err = klient.Universal.TykOAS().Create(ctx, apiID, tykOASDoc); err != nil {
 			log.Error(err, "failed to create Tyk OAS API")
-			return "", err
+			return "", "", err
 		}
 	} else {
+		oasApiDefOnTyk, _ := klient.Universal.TykOAS().Get(ctx, apiID) //nolint
+		// If we have same OAS API Definition on Tyk, we do not need to send Update to Tyk.
+		// So, we can simply return to main reconciliation logic.
+		if isSame(tykOASCrd.Status.LatestTykSpecHash, oasApiDefOnTyk) &&
```
[the remark this answers is not in the document]
Comment (buraksekili) [reply controllers/tykoasapidefinition_controller.go]: updated accordingly in 1eefac239ad6a28b1e1cdc1cab68f40e39f99d96. Could you please take a look?

File: api/v1alpha1/tykoasapidefinition_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (komalsukhani) [reply api/v1alpha1/tykoasapidefinition_types.go:83]: When I read it for the first time I thought it's CRD Spec Hash. But later I realized it's not.
That's why I added this comment. 
It's okay to keep it as is. It was just first impression.

**Review by komalsukhani — APPROVED** [review]

**Comment by bojantyk** [comment]
LGTM , @buraksekili  we can proceed with merging this PR. 
Thanks,

**Review by bojantyk — APPROVED** [review]

**Pull request merged**